### PR TITLE
Week1

### PR DESCRIPTION
**Pull request by hvariant**
latecomer catching up with homework

### CODE REVIEW

File: cis194/week1/zhansongl/credit_card.hs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    \xs -> (sumDigits xs) == (sum . numDigits $ strDigits xs)
+  where strDigits xs = foldr (++) "" $ map show xs
+        numDigits s
+          | s == "" = [0]
+          | otherwise = toDigits $ read s
+
+validate :: Integer -> Bool
+validate n = ((sumDigits . doubleEveryOther $ toDigits n) `mod` 10) == 0
+
+main :: IO ()
```
Comment (shineli1984) [inline cis194/week1/zhansongl/credit_card.hs]: Nit: Usually we are putting tests in a separate spec module.

File: cis194/week1/zhansongl/credit_card.hs [hunk 2]
```diff
@@ -0,0 +1,125 @@
+import Test.QuickCheck
+import Test.Hspec
+
+toDigitsRev :: Integer -> [Integer]
+toDigitsRev n
+  | n >= 0 = digit n
+  | otherwise = []
+  where digit 0 = []
+        digit n = n `mod` 10 : digit (n `div` 10)
```
Comment (shineli1984) [inline cis194/week1/zhansongl/credit_card.hs]: For positive integer, rem is preferred for performance reason.

File: cis194/week1/zhansongl/credit_card.hs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  where types = n::Integer
+        joinDigits xs = foldr (++) "" $ map show xs
+        showDigits n
+          | n <= 0 = "" -- weird restriction, but okay
+          | otherwise = show n
+
+doubleEveryOtherRev :: [Integer] -> [Integer]
+doubleEveryOtherRev [] = []
+doubleEveryOtherRev [x] = [x]
+doubleEveryOtherRev (x:y:xs) = (x:2*y:doubleEveryOtherRev xs)
```
Comment (shineli1984) [inline cis194/week1/zhansongl/credit_card.hs]: nit: Usually, we put spaces before and after `:`. For some codebase, we even put space after `(` and before `)`.

```suggestion
doubleEveryOtherRev ( x : y : xs ) = ( x : 2 * y : doubleEveryOtherRev xs )
```

File: cis194/week1/zhansongl/credit_card.hs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        padList (x:xs) = (x:0:padList xs)
+
+doubleEveryOtherRev_prop3 xs = (sum xs) == sum (doubleEveryOtherRev $ padList xs)
+  where types = xs::[Integer]
+        padList [] = []
+        padList [x] = [x]
+        padList (x:xs) = (x:0:padList xs)
+
+doubleEveryOther :: [Integer] -> [Integer]
+doubleEveryOther = reverse . doubleEveryOtherRev . reverse
```
Comment (shineli1984) [inline cis194/week1/zhansongl/credit_card.hs]: Given we are going to sum the digits after double every other digits from the right, do we need to reverse the list back?

**Comment by shineli1984** [comment]
I'll merge this for now and if you want to address the comments, you can create another PR.

File: cis194/week1/zhansongl/credit_card.hs (hunk 1 above)
Comment (hvariant) [reply cis194/week1/zhansongl/credit_card.hs]: will fix in a separate PR

File: cis194/week1/zhansongl/credit_card.hs (hunk 2 above)
Comment (hvariant) [reply cis194/week1/zhansongl/credit_card.hs]: hmm, didn't know about this, will fix

File: cis194/week1/zhansongl/credit_card.hs (hunk 3 above)
Comment (hvariant) [reply cis194/week1/zhansongl/credit_card.hs]: will fix

File: cis194/week1/zhansongl/credit_card.hs (hunk 4 above)
Comment (hvariant) [reply cis194/week1/zhansongl/credit_card.hs]: Hmm, if the only goal is to finish `validate`, then I suppose there is no point in reversing the list back. For the sake of this exercise though, if the list is returned in reverse it doesn't satisfy the requirements.

**Pull request merged**